internal/api/lib.go

```diff
@@ -81,814 +113,753 @@ func InitCache(config types.VMConfig) (Cache, error) {
 	return Cache{ptr: ptr, lockfile: *lockfile}, nil
 }
 
+// logCleanupError logs errors that occur during cleanup operations.
+// These errors are not critical as cleanup will happen when the process exits anyway.
+func logCleanupError(op string, err error) {
+	// If printing the error fails, we don't care.
+	// We can't log it anywhere, as that might cause infinite loops.
+	//nolint:gocritic
+	_, _ = fmt.Fprintf(os.Stderr, "warning: %s: %v\n", op, err)
+}
+
+// ReleaseCache releases the resources associated with the cache.
 func ReleaseCache(cache Cache) {
+	// First close the lockfile to release the lock
+	err := cache.lockfile.Close()
+	if err != nil {
+		logCleanupError("failed to close lockfile", err)
+		return
+	}
+	// Only release the cache if the lockfile was closed successfully
 	C.release_cache(cache.ptr)
-
-	cache.lockfile.Close() // Also releases the file lock
 }
 
+// StoreCode stores the given wasm code in the cache.
 func StoreCode(cache Cache, wasm []byte, persist bool) ([]byte, error) {
-	w := makeView(wasm)
-	defer runtime.KeepAlive(wasm)
+	if wasm == nil {
+		return nil, errors.New("null/nil argument")
+	}
+
+	// Check the WASM validity
+	wasmErr := validateWasm(wasm)
+	if wasmErr != nil {
+		return nil, wasmErr
+	}
+
 	errmsg := uninitializedUnmanagedVector()
-	checksum, err := C.store_code(cache.ptr, w, cbool(true), cbool(persist), &errmsg)
-	if err != nil {
-		return nil, errorWithMessage(err, errmsg)
+	csafeVec := storeCodeSafe(cache.ptr, wasm, true, persist, &errmsg)
+	if csafeVec == nil {
+		// Get the error message from the Rust code
+		safeVec := CopyAndDestroyToSafeVector(errmsg)
+		errMsg := string(safeVec.ToBytesAndDestroy())
```

Comment:
None of the uses of `CopyAndDestroyToSafeVector` make sense here. You always just call `ToBytesAndDestroy` immediately anyways. In that case, you might as well just use the existing `copyAndDestroyUnmanagedVector` directly, which returns perfectly safe `[]byte` to you.
Using `CopyAndDestroyToSafeVector` gives no additional safety at all. Nothing stops you from calling it twice and therefore double-freeing.

Reply:
I was worried about this.